### ISSUE TITLE
support `gradient` for extrapolation with constant filling

### ISSUE THREAD

**Issue by johnnychen94** [JuliaMath/Interpolations.jl#438]
Separated from the discussion in #434 and this is clearly an Interpolations issue.

When the extrapolation is constructed using a constant value, the out-of-range gradient is clearly `0`.

```julia
x = [1, 2, 3]
itp = interpolate(x, BSpline(Linear()))
etp = extrapolate(itp, 0)
etp(-1)
Interpolations.gradient(etp, -1) # not works
```

```julia
julia> Interpolations.gradient(etp, -1)
ERROR: MethodError: no method matching getfirst(::Int64)
Closest candidates are:
  getfirst(::Interpolations.Flag) at /Users/jc/Documents/Julia/Interpolations.jl/src/utils.jl:17
  getfirst(::Tuple) at /Users/jc/Documents/Julia/Interpolations.jl/src/utils.jl:18
Stacktrace:
 [1] _tcollect
   @ ~/Documents/Julia/Interpolations.jl/src/utils.jl:23 [inlined]
 [2] tcollect(f::typeof(Interpolations.etpflag), itp::Interpolations.FilledExtrapolation{Float64, 1, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear}, Tuple{Base.OneTo{Int64}}}, BSpline{Linear}, Int64})
   @ Interpolations ~/Documents/Julia/Interpolations.jl/src/utils.jl:22
 [3] gradient(etp::Interpolations.FilledExtrapolation{Float64, 1, Interpolations.BSplineInterpolation{Float64, 1, Vector{Float64}, BSpline{Linear}, Tuple{Base.OneTo{Int64}}}, BSpline{Linear}, Int64}, x::Int64)
   @ Interpolations ~/Documents/Julia/Interpolations.jl/src/extrapolation/extrapolation.jl:68
 [4] top-level scope
   @ REPL[67]:1
```

cc: @SomTambe @DhairyaLGandhi

**Comment by SomTambe** [comment]
So I tackled it like this - 
```julia
struct ZeroRGB{GT <: Union{Nothing, Interpolations.GridType}} <: Interpolations.BoundaryCondition
    gt::GT
end

julia> function Interpolations.inbounds_index((flagl,flagu)::Tuple{Nothing,Union{Flat,Line,ZeroRGB}}, (l,u), x, etp, xN)
           Interpolations.minp(x,u)
       end

julia> function Interpolations.inbounds_index((flagl,flagu)::Tuple{Union{Flat,Line,ZeroRGB},Flag}, (l,u), x, etp, xN)
           inbounds_index((nothing,flagu), (l,u), Interpolations.maxp(l,x), etp, xN)
       end

julia> Interpolations.extrapolate_gradient(::ZeroRGB, x, xs, g) = ifelse(x==xs, g, zero(g))

julia> etp2 = extrapolate(itp2, ZeroRGB())
5-element extrapolate(interpolate(::Array{RGB{Float64},1}, BSpline(Linear())), ZeroRGB()) with element type RGB{Float64}:
 RGB{Float64}(0.8144801416777425,0.4757000270645231,0.15316553311379133)
 RGB{Float64}(0.6425186760583639,0.03906742921171458,0.32000625630081836)
 RGB{Float64}(0.28970825978505066,0.9861789926757722,0.43937870045708927)
 RGB{Float64}(0.30136717766901566,0.47621620936904274,0.656569552445307)
 RGB{Float64}(0.23334204215592846,0.06232701866610957,0.15414915532875062)

julia> Interpolations.gradient(etp2, 6)
1-element SVector{1, RGB{Float64}} with indices SOneTo(1):
 RGB{Float64}(0.0,0.0,0.0)
```

**Comment by johnnychen94** [comment]
I think we only need an implementation for `gradient(etp::FilledExtrapolation, x::Vararg{Number, N})` method...

**Comment by SomTambe** [comment]
We can do this for general cases - 
```julia
julia> function Interpolations.gradient(etp::Interpolations.FilledExtrapolation{T, N, ITP, IT, FT}, xs::Vararg{Number, N}) where {T, N, ITP, IT<:Union{Tuple{Vararg{Interpolations.InterpolationType, N}}, Interpolations.InterpolationType}, FT}
           e = eltype(etp.itp)
           ptype = promote_type(e, typeof(etp.fillvalue))
           return SVector{1, ptype}(zero(ptype))
       end

julia> etp2
5-element extrapolate(interpolate(::Array{RGB{Float64},1}, BSpline(Linear())), RGB{Float64}(0.0,0.0,0.0)) with element type RGB{Float64}:
 RGB{Float64}(0.825570693135897,0.9327818758605733,0.8298271227499754)
 RGB{Float64}(0.47833422796307645,0.3123829559320115,0.8876717902707263)
 RGB{Float64}(0.47509547952726705,0.7576030332649815,0.19032958769526287)
 RGB{Float64}(0.47263407308494365,0.8698957806775736,0.6454195705271191)
 RGB{Float64}(0.6341816805190843,0.8003466512785806,0.34013593030636424)

julia> Interpolations.gradient(etp2, 6)
1-element SVector{1, RGB{Float64}} with indices SOneTo(1):
 RGB{Float64}(0.0,0.0,0.0)

julia> etp
3×3 extrapolate(interpolate(::Matrix{Float64}, BSpline(Linear())), 55) with element type Float64:
 0.58301   0.113481   0.479648
 0.126568  0.0497296  0.214082
 0.822931  0.244805   0.441772

julia> Interpolations.gradient(etp, 10, 10)
hello
1-element SVector{1, Float64} with indices SOneTo(1):
 0.0
```
What say?

**Comment by johnnychen94** [comment]
The gradient is only zero for out-of-range indices, for in-range indices we should fall back to interpolate gradient. Hence it should be something like this:

```julia
function Interpolations.gradient(etp::Interpolations.FilledExtrapolation{T}, xs::Vararg{Number, N}) where {T, N}
    itp = parent(etp)
    if checkbounds(Bool, itp, x...)
        gradient(itp, x...)
    else
        return SVector(zero(T))
    end
end
```

It might not be `zero(T)` here but the point is to keep the output type the same as `gradient(itp, x...)` for type stability purpose.

**Comment by SomTambe** [comment]
Yes, that was silly of me. I'll make the changes accordingly. Should I make a pull request for the same?

**Comment by johnnychen94** [comment]
Yes, a PR would be good for review.

**Comment by johnnychen94** [comment]
closed by #439
